refactor(compiler): Add conditionalCreate instruction

thePunderWoman opened the pull request:
This adds a new instruction for dealing with creating conditionals. It ensures flags are set on the TNode for later identification during hydration.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

thePunderWoman committed:
refactor(compiler): Add conditionalCreate instruction

This adds a new instruction for dealing with creating conditionals. It ensures flags are set on the TNode for later identification during hydration.

AndrewKushnir commented on packages/core/src/render3/instructions/control_flow.ts:
```diff
@@ -43,6 +43,67 @@ import {
+    attrs,
+    localRefsIndex,
+    localRefExtractor,
+    [TNodeFlags.isControlFlowStart, TNodeFlags.isInControlFlow],
```
nit: we can avoid creating an array here and just combine the flags here (similarly for the other instruction):
```suggestion
    TNodeFlags.isControlFlowStart | TNodeFlags.isInControlFlow,
```

AndrewKushnir commented on packages/core/src/render3/instructions/control_flow.ts:
```diff
@@ -43,6 +43,67 @@ import {
   removeLViewFromLContainer,
 } from '../view/container';
 
+export function ɵɵconditionalCreate(
```
Similar to other instructions, we can add some docs and the `@codeGenApi` annotation:
```suggestion
/**
 * ...
 * @codeGenApi
 */
export function ɵɵconditionalCreate(
```

thePunderWoman committed:
fix(core): fixes timing of hydration cleanup on control flow

This properly cleans up stale control flow branches in the case
that branches change between server and client at the same
timing as NgIf / NgSwitch.

fixes: #58670
fixes: #60218

crisbeto reviewed (approved):
LGTM for the compiler parts.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/chaining.ts:
```diff
@@ -100,7 +106,12 @@ function chainOperationsInList(opList: ir.OpList<ir.CreateOp | ir.UpdateOp>): vo
-    if (chain !== null && chain.instruction === instruction && chain.length < MAX_CHAIN_LENGTH) {
+    if (
+      chain !== null &&
+      (chain.instruction === instruction ||
```
The `||` here means that `conditionalCreate` will be considered for chaining with `conditionalCreate`, which is not valid.

I would suggest replacing `CHAINABLE` completely with `CHAIN_COMPATIBILITY`:
* the `CHAINABLE.has` check becomes `CHAIN_COMPATIBILITY.has` (keys are chain starts)
* the `chain.instruction === instruction` becomes `CHAIN_COMPATIBILITY.get(chain.instruction) === instruction` (it's legal to chain `instruction` to the chain's root).

In most cases, `CHAIN_COMPATIBILITY` will map instructions to themselves, except for the conditional ones.

AndrewKushnir commented on packages/core/src/hydration/cleanup.ts:
```diff
@@ -161,3 +162,66 @@ export function cleanupHydratedDeferBlocks(
+/**
+ * In the case that control flow has taken a different branch between server
+ * and client, we need to find and remove the now stale dehydrated branch.
+ *
+ * @param hostLView
+ * @param tNode
```
```suggestion
```

AndrewKushnir commented on packages/core/src/render3/instructions/control_flow.ts:
```diff
@@ -438,6 +563,20 @@ export function ɵɵrepeater(collection: Iterable<unknown> | undefined | null):
+          if (
+            emptyTemplateTNode.tView!.firstUpdatePass &&
+            emptyTemplateTNode.tView!.ssrId !== null &&
+            !foundDehydratedView
```
It looks like we compute the value of the `foundDehydratedView` in a different branch above and it doesn't seem to affect this call site.

AndrewKushnir commented on packages/core/src/render3/instructions/control_flow.ts:
```diff
@@ -86,6 +179,19 @@ export function ɵɵconditional<T>(matchingTemplateIndex: number, contextValue?:
+          templateTNode.tView!.ssrId !== null &&
+          dehydratedView === null
+        ) {
+          removeStaleDehydratedBranch(hostLView, templateTNode);
```
I feel like we should move this logic into the `findMatchingDehydratedView`, which currently contains a code path to cleanup dehydrated views for the `*ngFor` and `*ngIf`. In this case, the code would be tree-shakable without extra code switches (in the `api.ts`) and the cleanup paths for old and new control flows would co-exist in one function.

thePunderWoman replied:
I feel like this would really complicate the logic of `findMatchingDehydratedView`, which is already side-effectful since it slices the dehydrated view array. IMHO this would need a larger refactor to make it clean and easier to follow. I do like that right now it is a separate code path that is clear what it does. So I'll take a look, but maybe this is something we can do in a follow up.

thePunderWoman replied:
OK, found a solution that creates a wrapper that does this work.

thePunderWoman commented on packages/core/src/render3/instructions/control_flow.ts:
```diff
@@ -58,12 +156,7 @@ export function ɵɵconditional<T>(matchingTemplateIndex: number, contextValue?:
-    prevMatchingTemplateIndex !== -1
-      ? getLContainer(hostLView, HEADER_OFFSET + prevMatchingTemplateIndex)
-      : undefined;
+  const prevContainer = getPrevContainer(hostLView, bindingIndex);
```
I originally moved it to reuse it elsewhere, but then that didn't need to happen. So I'll restore this instead.

AndrewKushnir commented on packages/core/src/hydration/views.ts:
```diff
@@ -61,6 +64,14 @@ export function locateDehydratedViewsInContainer(
+ * stored on a given lContainer.
+ * Returns `null` by default, when hydration is not enabled.
+ */
+let _findAndReconcileMatchingDehydratedViewsImpl: typeof findAndReconcileMatchingDehydratedViewsImpl =
```
It'd be great to collect all functions that participate in this code switch in a single place (similar to how `hasMatchingDehydratedView` switch is implemented above). Currently those functions are separated by other functions in this file.

thePunderWoman commented on packages/core/src/hydration/utils.ts:
```diff
@@ -759,3 +760,32 @@ export function verifySsrContentsIntegrity(doc: Document): void {
+ */
+let _hasMatchingDehydratedViewImpl: typeof hasMatchingDehydratedViewImpl = () => false;
+
+function hasMatchingDehydratedViewImpl(lContainer: LContainer, template: string | null): boolean {
```
Yeah, I was pretty sure I removed that, but must have left some of it behind.

thePunderWoman commented on packages/core/src/hydration/utils.ts:
```diff
@@ -759,3 +760,32 @@ export function verifySsrContentsIntegrity(doc: Document): void {
+  // Verify whether the first dehydrated view in the container matches
+  // the template id passed to this function (that originated from a TView
+  // that was used to create an instance of an embedded or component views.
+  return views[0].data[TEMPLATE_ID] === template;
```
That is because this actually _is_ the same function and this instance shouldn't be here anymore. I just forgot to remove it. The one in `views.ts` is the correct one.

thePunderWoman commented on packages/core/src/hydration/views.ts:
```diff
@@ -95,8 +106,38 @@ function findMatchingDehydratedViewImpl(
+  // Verify whether the first dehydrated view in the container matches
+  // the template id passed to this function (that originated from a TView
+  // that was used to create an instance of an embedded or component views.
+  return views[0].data[TEMPLATE_ID] === template;
```
This is the correct instance. Thanks for catching this.

thePunderWoman commented:
Caretaker note: This needs cl/739254495 patched in when it lands.

AndrewKushnir commented on packages/core/src/hydration/views.ts:
```diff
@@ -95,13 +112,101 @@ function findMatchingDehydratedViewImpl(
+  }
+}
+
+function hasMatchingDehydratedView(lContainer: LContainer, template: string | null): boolean {
```
IIUC, this function can be used within the `findMatchingDehydratedViewImpl` one here: https://github.com/angular/angular/blob/main/packages/core/src/hydration/views.ts#L77-L85

alxhub commented:
This PR was merged into the repository by commit 13d1c8ab38707b06dbed4941a556f982665b3304.

The changes were merged into the following branches: main
